keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,617 @@
+### Caching remaining capacity via the API server
+
+CSI defines the GetCapacity RPC call for the controller service to
+report information about the current capacity but it's not accessible
+in the Kubernetes scheduler. The existing `CSIDriver` type gets
+extended with a `Status` field that (initially) only holds the
+capacity information in a format that is sufficiently flexible to
+handle the use cases outlined above. In the future, that `Status`
+field could also be extended to represent other dynamic information
+about the driver.
+
+The CSI driver deployment is already responsible for creating a
+`CSIDriver` object for itself. Now it will also have to enable and
+configure the adding, updating and removing of capacity
+information. To simplify that, the `external-provisioner` will be
+extended to handle this based on the already existing APIs (i.e. no
+changes needed in a CSI driver itself).
+
+The Kubernetes scheduler monitors `CSIDriver` objects and uses that
+information when it comes to making scheduling decisions. If the
+driver does not indicate that it supports capacity reporting, then the
+scheduler proceeds just as it does now, so nothing changes for
+existing CSI driver deployments.
+
+### API
+
+All Kubernetes API changes are under the `CSIStorageCapacity` feature
+gate.
+
+#### Storage capacity
+
+The prefix “CSI” is used to avoid confusion with other types that have
+a similar name and because some aspects are specific to CSI.
+
+```
+type CSIDriver struct {
+    ...
+
+	// Specification of the CSI Driver.
+	Spec CSIDriverSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// Status of the CSI Driver.
+	// +optional
+	Status CSIDriverStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+type CSIDriverSpec struct {
+    ...
+
+	// CapacityTracking defines whether the driver deployment will provide
+	// capacity information as part of the driver status.
+	// +optional
+	CapacityTracking *bool `json:"capacityTracking,omitempty" protobuf:"bytes,4,opt,name=capacityTracking"`
+}
+
+// CSIDriverStatus represents dynamic information about the driver and
+// the storage provided by it, like for example current capacity.
+type CSIDriverStatus struct {
+	// Each driver can provide access to different storage pools
+	// (= subsets of the overall storage with certain shared
+	// attributes).
+	//
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	Storage []CSIStoragePool `patchStrategy:"merge" patchMergeKey:"name" json:"storage,omitempty" protobuf:"bytes,1,opt,name=storage"`
+}
+
+// CSIStoragePool identifies one particular storage pool and
+// stores the corresponding attributes.
+//
+// A pool might only be accessible from a subset of the nodes in the
+// cluster. That subset can be identified either via NodeTopology or
+// NodeList, but not both. If neither is set, the pool is assumed
+// to be available in the entire cluster.
+type CSIStoragePool struct {
```

Comment:
For local storage in a very large cluster, is it too long for users to read the information in one API ?  It seems we need to have a CSIStoragePool struct for each node.
And also, is it reasonable to organize the storage pool by storage class ?  I agree it is good for network storage, but it seems not that good for local storage .

Reply:
> For local storage in a very large cluster, is it too long for users to read the information in one API object ?

That may become an issue, yes.

> It seems we need to have a CSIStoragePool struct for each node.

But that's then not meant to be readable by users, is it? As discussed in https://github.com/kubernetes/enhancements/pull/1347#discussion_r342008418, the name of those objects most likely will have to be generated such that they don't mean anything.

What about storing the information as proposed in this KEP, but then not show it in `kubectl describe`? `kubectl describe pods` does that for `spec.affinity`. It's still available to users when using `kubectl get -o yaml`. I'm not sure how that is implemented and whether it is a good idea, but I at least wanted to bring up... :man_shrugging: 

> And also, is it reasonable to organize the storage pool by storage class ? I agree it is good for network storage, but it seems not that good for local storage .

I don't follow here. Are you saying that local storage can ignore storage classes? That's not necessarily the case. The example where available usable storage depends on storage class parameters is LVM with striped vs. mirrored mode selected via the storage class. That is local storage.

Reply:
>But that's then not meant to be readable by users, is it?

Yes, exactly.

>What about storing the information as proposed in this KEP, but then not show it in kubectl describe? kubectl describe pods does that for spec.affinity. It's still available to users when using kubectl get -o yaml

I guess users may want to know the topology info just like `capacity` and `allocatable` in node status.

>Are you saying that local storage can ignore storage classes? That's not necessarily the case. The example where available usable storage depends on storage class parameters is LVM with striped vs. mirrored mode selected via the storage class. That is local storage.

Yeah, actually, for local storage,  we just need one storage class with some necessary parameters in the cluster. If so, as proposed here, each node with local storage is a storage pool.... Their storageclass may be all the same. I am not sure if it is reasonable, just feel weird. 
It seems StorageClass and StoragePool does not have strict mapping relation. Why do we need to add StorageClass in StoragePool object.

Reply:
Or could we limit the StoragePool more strictly, and make it 1:1 mapping relationship with StorageClass ?

Reply:
> I guess users may want to know the topology info just like capacity and allocatable in node status.

So you are arguing that we should put storage information into the `Node.Status`? That only makes sense for one particular use case (local storage). I don't think we should develop a special solution just for that.

> Yeah, actually, for local storage, we just need one storage class with some necessary parameters in the cluster.

That may be the case for some drivers, but not all. We need a solution that is flexible enough to cover different scenarios. A driver with multiple different storage classes and capacity that depends on the storage class is one of those scenarios.

If a driver does not need that flexibility, it can produce information with just `storageClassName: <fallback>`, as in the "local storage" example:
```
status:
  storage:
  - classes:
    - capacity: 256G
      storageClassName: <fallback>
    name: node-1
    nodeList:
    - node-1
```

> Or could we limit the StoragePool more strictly, and make it 1:1 mapping relationship with StorageClass ?

Isn't that the special case above?

Reply:
> So you are arguing that we should put storage information into the Node.Status? That only makes sense for one particular use case (local storage). I don't think we should develop a special solution just for that.

No , i do not mean to put storage info to Node status.  What i want to propose is: create separate `StoragePool` objects and do not put them to `CSIDriver` so that it can be more easier for us to read.

>That may be the case for some drivers, but not all. We need a solution that is flexible enough to cover different scenarios. A driver with multiple different storage classes and capacity that depends on the storage class is one of those scenarios.

I agree local storage is not general enough. But it is the case we need to consider too. Putting all storage info together is not that user-friendly.

Reply:
> What i want to propose is: create separate StoragePool objects and do not put them to CSIDriver so that it can be more easier for us to look up to.

I'm not against that if people really see that as easier to use. For code working with these objects both works, because the information is the same.

But I am not convinced that this is really easier. Can you give an example of an operation that you would like to do and how that would look like if we had a StoragePool as stand-alone object? You can use the CRD from https://github.com/kubernetes/enhancements/pull/1347 as the API; it's a bit too limited right now (no support for storage classes), but in principle that's the alternative API.

If it helps, then I can also put a full API description for such a StoragePool into this KEPs' "alternatives" section.

Reply:
My point above is mostly for local storage.  Assuming that we have a 5k nodes cluster, and we want to manage the local storage Apps. As i mentioned above, it is hard for me to read the  local storage status in one API object, especially for complex situations, for example, one node with multiple VGs(LVM). It will be easy for me to look up to or manage them separately. Just like CSINode does (BTW: putting local storage to CSINode is ok for me too)
For network storage, i think it is also ok to have stand-alone StoragePool objects, each one represents the storage pool with specific topology limit.

Reply:
>  It will be easy for me to look up to or manage them separately.

This is where I have my doubts and therefore would like to hear more how you intend to do that in practice. For example, the "look up" part will probably not be a easy as you think because the name of the object you look for may just be `sp-ba603482-0c36-11ea-9b2b-27b0c0dd3fdd`. In some very special cases it might be possible to come up with a better name, but not in the general case.

> Just like CSINode does

There's a 1:1 mapping between node and `CSINode`, so the latter can use the node name as its own name. The mapping is much more complex for storage pools.

Reply:
I've updated the "alternatives" section: https://github.com/kubernetes/enhancements/blob/0401cb12601510ebb9848079f857e0823049ff24/keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md#separate-objects

@NickrenREN: please let me know whether this is what you had in mind and whether you still prefer it.

Reply:
Thanks, yes, considering local storage, that is what i wan.
BTW:  `Capacity` in the Status of CSIStoragePool here represents `the size of the largest volume that currently can be created.`   It should be managed by CSI driver, right ? It seems there is a race condition between scheduler and CSI driver. As you described,  it is just `a best-effort guess`.  
Why don't we make it represent the `Capacity` of the CSI Storage Pool ?

If so, maybe we need to change the unit from `*resource.Quantity` to `*resource.ResourceList`, because there may be multiple Capacities in one StoragePool, for example : One node has multiple VGs(LVM based).

Reply:
Thanks, yes, considering local storage, that is what i wan.
BTW:  `Capacity` in the Status of CSIStoragePool here represents `the size of the largest volume that currently can be created.`   It should be managed by CSI driver, right ? It seems there is a race condition between scheduler and CSI driver. As you described,  it is just `a best-effort guess`.  
Why don't we make it represent the `Capacity` of the CSI Storage Pool ?

If so, maybe we need to change the unit from `*resource.Quantity` to `*resource.ResourceList`, because there may be multiple Capacities in one StoragePool, for example : One node has multiple VGs(LVM based).

Reply:
>Can you describe what kind of operations you want to do with the data and how you would do them?

Sure, we have a 5k nodes kubernetes cluster which is sharded by specific labels. And Admins/Users need to care about local storage (Capacity & Allocatable). But since the cluster is sharded and Apps can only be deployed among one specific partition. In most cases, we do not need to look up to all nodes' local storage condition. Instead, several nodes in one partition is enough,  so it will be more clear for us to store the local storage separately.
Since our cluster is quite large, we care about the performance a lot. As i mentioned above, we sharded the cluster, and we make the scheduler work in parallel too.  Separate StoragePool Objects are more efficient and suitable for sharded schedulers too since one CSIDriver object may be a bottleneck.

Reply:
> Especially user-friendly when using kubectl describe...

But how does the user find the right object to call `kubectl describe` for? How user-friendly is that part going to be?

> Since our cluster is quite large, we care about the performance a lot.

That part is indeed crucial. But it also depends on the usage pattern of the performance-critical parts. So let's consider that aspect for the worst case, node-local storage. With the current proposal (`CSIDriver.Status`), there will be one producer per node, sending updates to it's part of the status via a patch operation. There's one consumer (the scheduler). It should better not create a watch because otherwise it'll constantly get updates even when there's nothing to schedule. Instead, it can retrieve the `CSIDriver` each time it is needed. I'm assuming here that scheduling pods with storage is going to be rare compared to scheduling pods without storage.

With `CSIStoragePool`, the writers will send a complete `Status` update for one object. There's still only one consumer (or two in your case?). That scheduler now must retrieve all objects with a certain `CSIStoragePool.Spec.DriverName` value each time it wants to schedule a pod with a volume using that driver. Because individual objects are smaller, a watch might make more sense here, although the argument about getting updates when they aren't needed still holds.

Does that sound right?

It's not obvious to me which approach will lead to better performance. I was hoping that some apiserver experts have already measured something like this and could share their experience. If that doesn't happen, then I have to proceed with prototyping both APIs and writing some microbenchmark to get some real numbers.

Reply:
>But how does the user find the right object to call kubectl describe for? How user-friendly is that part going to be?

For local storage, its (CSIStoragePool) name can be set to node name

>Does that sound right?

hmm, what i want to clarify here is:  We modify the k8s scheduler part. We create several schedulers and make them work in parallel. Each scheduler is only responsible for one partition (pods with specific NodeSelector and partial nodes with specific label). And schedulers can not share resources(nodes).
If we store all local storage info in one object, all schedulers need to look up to it, therefore, we need a lock or something like that. But if we create separate objects, each scheduler only care about partial CSIStoragePool objects, since scheduelers can not share nodes resource,  they can work in parallel and do not need locks.

I know the solution is not general enough. Schedulers will have to share some other informations. But at least for local storage, it is ok. For network storage, we can shard the storage pools manually and make it fit for multiple schedulers too.

Reply:
> For local storage, its (CSIStoragePool) name can be set to node name

This only works under the assumption that only one CSI driver is providing capacity information for that node. But indeed, if this assumption holds, then individual objects may work better. Thanks for clarifying this.

Reply:
I spoke with Jordan and he agrees that if the objects will end up scaling with the number of nodes, then it is preferable to have a separate object.  The main issues with a single object:
- any update to that single object, no matter how small, ends up sending the whole API object to all the watchers. This will not scale in large clusters.
- when an informer gets and update, it gets the whole object, you can't see only the small delta that has changed, so you end up needing to iterate through the entire object to process a small update.

Separating it out into separate objects means more work on the consumer end, especially to build up caches, but it substantially reduces API server load in large clusters.

Reply:
Both of these points apply to a usage pattern that does not necessarily apply for the implementation outlined in this KEP. But as @NickrenREN explained, they might want to use the API in a different way where they *do* apply, so I'm fine with switching to a individual objects. I'll update the KEP accordingly.